storage/src/test/java/tech/pegasys/teku/storage/server/kvstore/DatabaseTest.java

```diff
@@ -1666,6 +1666,40 @@ public void shouldStoreFinalizedState(final DatabaseContext context) throws IOEx
     assertThat(database.getLatestAvailableFinalizedState(ONE)).contains(beaconState);
   }
 
+  @TestTemplate
+  public void shouldStoreFinalizedStateAndRoots(final DatabaseContext context) throws IOException {
+    createStorageSystem(context, StateStorageMode.ARCHIVE, StoreConfig.createDefault(), false);
+
+    final BeaconState state0 = dataStructureUtil.randomBeaconState(ZERO);
+    final BeaconState state1 = dataStructureUtil.randomBeaconState(ONE);
+
+    final Bytes32 blockRoot0 = dataStructureUtil.randomBytes32();
+    final Bytes32 blockRoot1 = dataStructureUtil.randomBytes32();
+
+    assertThat(database.getLatestAvailableFinalizedState(ONE)).isEmpty();
+    database.storeFinalizedState(state0, blockRoot0);
+    database.storeFinalizedState(state1, blockRoot1);
+
+    assertThat(database.getLatestAvailableFinalizedState(ZERO)).contains(state0);
+    assertThat(database.getLatestAvailableFinalizedState(ONE)).contains(state1);
+
+    try (final Stream<Map.Entry<Bytes32, UInt64>> roots = database.getFinalizedStateRoots()) {
+      assertThat(getFinalizedStateRootsAtSlot(roots, ZERO)).contains(state0.hashTreeRoot());
+    }
+
+    try (final Stream<Map.Entry<Bytes32, UInt64>> roots = database.getFinalizedStateRoots()) {
+      assertThat(getFinalizedStateRootsAtSlot(roots, ONE)).contains(state1.hashTreeRoot());
+    }
```

Comment:
maybe it's better to just get the entries at slot 0 and 1 and compare to a list of map entries and cover both? may be easier to see...